itk_cucim/filtering/distance_map.py

```diff
@@ -0,0 +1,158 @@
+        for computing exact Euclidean distance transforms of binary images in
+        arbitrary dimensions," in IEEE Transactions on Pattern Analysis and
+        Machine Intelligence, vol. 25, no. 2, pp. 265-270, Feb. 2003.
+        :DOI:`10.1109/TPAMI.2003.1177156`
+    .. [2] Thanh-Tung Cao, Ke Tang, Anis Mohamed, and Tiow-Seng Tan. 2010.
+        Parallel Banding Algorithm to compute exact distance transform with the
+        GPU. In Proceedings of the 2010 ACM SIGGRAPH symposium on Interactive
+        3D Graphics and Games (I3D ’10). Association for Computing Machinery,
+        New York, NY, USA, 83–90.
+        :DOI:`10.1145/1730804.1730818`
+    .. [3] https://www.comp.nus.edu.sg/~tants/pba.html
+    """
+    if image.dtype == np.uint8:
+        # can avoid copy from uint8->bool
+        image = image.view(bool)
+    else:
+        # copy=False to omit copy of images that are already boolean
+        image = image.astype(bool, copy=False)
+
+
+    footprint = cp.ones((3, 3), dtype=bool)
+    if inside_is_positive:
+        image_in = ~image
+    else:
+        # boundary: erode by one pixel to get an equivalent result to ITK
+        image_in = binary_erosion(image, footprint=footprint)
+
+    distance_kwargs = dict(return_distances=True, return_indices=False)
+    if spacing is not None:
+        spacing = tuple(spacing)
+        if any(s != 1.0 for s in spacing):
+            distance_kwargs['sampling'] = spacing
+
+    # distance transform of the eroded image
+    distance = distance_transform_edt(image_in, **distance_kwargs)
+
+    # now compute a second unsigned distance transform
+    if inside_is_positive:
+        # boundary: erode by one pixel to get an equivalent result to ITK
+        image_in = binary_erosion(image, footprint=footprint)
+    else:
+        image_in = ~image
+    distances_inv = distance_transform_edt(image_in, **distance_kwargs)
+
+    if squared_distance:
+        distances_inv *= distances_inv
+        distance *= distance
+
+    # subtract the unsigned transforms to get the signed result
+    distances_inv -= distance
+    return distances_inv
+
+
+@helpers.accept_array_like_xarray_torch
+def cucim_signed_maurer_distance_map_image_filter(*args, **kwargs):
+    input_image = args[0]
+
+    # TODO: remove requirement to cast to float32
+    #   Output dtype is float32 for the distance transform.
+    #   PyImageFilter currently supports 2D, 3D or 4D images of dtypes:
+    #     float32, float64, uint8, uint16, int16
+    #   with restriction that input and output dtypes must match.
+    #   To get float32 output, we must cast the input to float32 here.
+    float32_type = itk.Image[itk.F, input_image.ndim]
+    input_image = itk.cast_image_filter(
+        input_image, ttype=(type(input_image), float32_type)
+    )
+    args = (input_image, ) + args[1:]
+
+    ref_filt = itk.SignedMaurerDistanceMapImageFilter.New(*args, **kwargs)
+    wrapper = itk.PyImageFilter.New(input_image)
```

Comment:
Any ideas how to avoid making a copy here? As far as I could tell `PyImageFilter` bindings only exist for cases where the output dtype that matches the input dtype. However, in the case of the distance filter, the output is in floating point even for integer valued inputs.

Reply:
Existing result of `itk.PyImageFilter.GetTypes()` is
```
  [<class 'itk.itkImagePython.itkImageD2'>, <class 'itk.itkImagePython.itkImageD2'>]
  [<class 'itk.itkImagePython.itkImageD3'>, <class 'itk.itkImagePython.itkImageD3'>]
  [<class 'itk.itkImagePython.itkImageD4'>, <class 'itk.itkImagePython.itkImageD4'>]
  [<class 'itk.itkImagePython.itkImageF2'>, <class 'itk.itkImagePython.itkImageF2'>]
  [<class 'itk.itkImagePython.itkImageF3'>, <class 'itk.itkImagePython.itkImageF3'>]
  [<class 'itk.itkImagePython.itkImageF4'>, <class 'itk.itkImagePython.itkImageF4'>]
  [<class 'itk.itkImagePython.itkImageSS2'>, <class 'itk.itkImagePython.itkImageSS2'>]
  [<class 'itk.itkImagePython.itkImageSS3'>, <class 'itk.itkImagePython.itkImageSS3'>]
  [<class 'itk.itkImagePython.itkImageSS4'>, <class 'itk.itkImagePython.itkImageSS4'>]
  [<class 'itk.itkImagePython.itkImageUC2'>, <class 'itk.itkImagePython.itkImageUC2'>]
  [<class 'itk.itkImagePython.itkImageUC3'>, <class 'itk.itkImagePython.itkImageUC3'>]
  [<class 'itk.itkImagePython.itkImageUC4'>, <class 'itk.itkImagePython.itkImageUC4'>]
  [<class 'itk.itkImagePython.itkImageUS2'>, <class 'itk.itkImagePython.itkImageUS2'>]
  [<class 'itk.itkImagePython.itkImageUS3'>, <class 'itk.itkImagePython.itkImageUS3'>]
  [<class 'itk.itkImagePython.itkImageUS4'>, <class 'itk.itkImagePython.itkImageUS4'>]
```

For the purposes of this function it would be useful to add
```
[<class 'itk.itkImagePython.itkImageUC2'>, <class 'itk.itkImagePython.itkImageF2'>]
[<class 'itk.itkImagePython.itkImageUC3'>, <class 'itk.itkImagePython.itkImageF3'>]
[<class 'itk.itkImagePython.itkImageUC4'>, <class 'itk.itkImagePython.itkImageF4'>]
[<class 'itk.itkImagePython.itkImageUS2'>, <class 'itk.itkImagePython.itkImageF2'>]
[<class 'itk.itkImagePython.itkImageUS3'>, <class 'itk.itkImagePython.itkImageF3'>]
[<class 'itk.itkImagePython.itkImageUS4'>, <class 'itk.itkImagePython.itkImageF4'>]
[<class 'itk.itkImagePython.itkImageSS2'>, <class 'itk.itkImagePython.itkImageF2'>]
[<class 'itk.itkImagePython.itkImageSS3'>, <class 'itk.itkImagePython.itkImageF3'>]
[<class 'itk.itkImagePython.itkImageSS4'>, <class 'itk.itkImagePython.itkImageF4'>]
```
(4D added for consistency with above, but we only have 2D and 3D implementations for this particular cuCIM function)

Reply:
Having those would allow avoiding the call to `cast_image_filter` and the corresponding copy back to uint8 on the GPU. It would also avoid host->device transfer overhead (due to larger size of float vs. 8 or 16-bit integers)